### ISSUE TITLE
AI has north-west bias (GNA #12978)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1145]
**[Original submission](https://gna.org/bugs/index.php?12978)** by anonymous on 2009-02-12

see: this forum thread

http://www.wesnoth.org/forum/viewtopic.php?p=327956#p327956

Basically, the AI has a tendency to, when all other options are equal, move to the NW-most hex available; this makes it drift in that direction on large maps, makes it excessively predictable, etc... it'd be nice to have it fixed.

(Reproduced on all)
Release: 1.5.9+svn
Priority: 5 - Normal
Severity: 2 - Minor

**Comment by wesnoth-bugs** [comment]
**Modified on 2009-02-16**

**esr wrote:**

> Old bug, still important.  See  #8410 (AI makes big mistakes in custom maps!)

**esr changed status:** None -> Duplicate

**Comment by wesnoth-bugs** [comment]
**Modified on 2009-02-23**

**esr closed the issue**

**Comment by wesnoth-bugs** [comment]
**Modified on 2009-03-06**

**esr wrote:**

> Reactivating, since #8410 is closed due to the other issues being judged subjective or nonreproducible by Dave -- he says this one is valid.  Regrading it minor.

**esr changed severity:** 3 - Normal -> 2 - Minor

**esr changed status:** Duplicate -> None

**Comment by wesnoth-bugs** [comment]
**Modified on 2009-03-06**

**esr reopened the issue**

**Comment by wesnoth-bugs** [comment]
**Modified on 2010-03-09**

**crab changed status:** None -> Postponed

**crab wrote:**

> (postponed till 1.9, when there'll be more time to investigate the issue)

**Comment by wesnoth-bugs** [comment]
**Modified on 2012-01-29**

**shadowmaster changed status:** Postponed -> None

**Comment by mattsc** [comment]
This is still a valid issue, but it is not likely to be fixed anytime soon.

**Comment by Wedge009** [comment]
I have no idea how the AI currently works, but is it a case of the hexes around a unit being labelled 1-6, and - when all options are equal - the AI picks the direction of hex 1? (Assuming hex 1 is NW.)

**Comment by GregoryLundberg** [comment]
It's probably 0 to 7 and defaulting to 0. The issue, it seems, would be to find it :P

**Comment by mattsc** [comment]
> I have no idea how the AI currently works, but is it a case of the hexes around a unit being labelled 1-6, and - when all options are equal - the AI picks the direction of hex 1?

Yes, that's exactly what's going on.  Specifically, hexes are considered in order, at least for some evaluations, and a later hex is only chosen if its rating is strictly greater than (as in, not equal to) the previous max rating.

It would be trivial to fix, e.g. simply by adding a very small random number to all final ratings (in fact, some of the Micro AIs do something like this, if I remember correctly).  I've been hesitant to do so because somebody a long time ago (might have been Crab, I don't remember) once told me that the AI behavior should be reproducible and not have random elements, although I never quite understood why (other than for trouble-shooting, of course).

**Comment by GregoryLundberg** [comment]
This is one case where it needs a random choice. There are probably others.

I'd take the comment about no randomness in the context of where there is a clear winner, that should be the choice; and it should always be the winner (no random factors deeper down).

The issue here is, when all choices have the same weight, the one which is chosen should be selected randomly, specifically to avoid cases like the NW bias being "gamable"

**Comment by soliton-** [comment]
If reproducability is needed for unit tests for example it shouldn't be too hard to make a switch to turn the randomness off at compile or run time.

**Comment by stevecotton** [comment]
A pseudo-random number, such as a hash of the game state, could be both reproducible and sufficiently like a random number to remove the bias.

**Comment by Wedge009** [comment]
@mattsc I understand you're reducing your focus on Wesnoth (perfectly understandable) - any pointers on where the hex 0-5 decision is made or was finding that part of the difficulty in resolving this issue?

**Comment by mattsc** [comment]
A good part of the challenge is determining all the places where this happens.  A couple obvious ones are

1. Recruiting: the AI starts with all hexes adjacent to the leader and goes through them in order lowest x value first, followed by lowest y value.  Assuming all else being equal (no enemy threats  to castle hexes etc.), the one with lowest x and y value, that is, the one in the NW is picked first.
2. Attacks: again all else being equal (terrain, enemy threat, own units support), the AI finds all hexes adjacent to an enemy and then goes through them in some predetermined order and picks the first it encounters.

[ That's at least how it used to be, things might have changed since I last looked into this.  So take these as examples of the principle of what happens, even if they might not be literally what is happening any more. Also, in some cases it appears to be lowest x followed by lowest y, while in others it is starting in the NW and going around counterclockwise. Depends on the order in which the underlying functions returns the hexes. ]

Similar things happen in other CAs when a find-multiple-hexes query returns hexes always in the same order, but I do not know all the places in the code where that is the case.

**Comment by Wedge009** [comment]
I see. I thought it was a case of just one or two locations where this hex indexing was occurring. No worries, at least we have an idea of what's happening. Thanks.

**Comment by mattsc** [comment]
I really don't know in how many places it is occurring.  Or maybe more importantly: in how many places it matters, because in practice the "all else being equal" might simply not happen.  But I do believe that all AI evaluation functions need to be checked if we decide to take care of this, and that's something I don't think anybody has ever done.
